### PR TITLE
confdgnmi

### PR DESCRIPTION
**Pull request by micnovak**
allow_aggregation, aggregation tests, tests for gnmi-tools.yang data model (only ConfD), Api tests renamed to ConfD tests, updated docs

### CODE REVIEW

File: tests/client_server_test_base.py [hunk 1]
```diff
@@ -237,9 +251,9 @@ def _test_get_subscribe(self, is_subscribe=False,
                                      if_id)]
         list_paths = [
             GrpcBase.mk_gnmi_if_path(self.list_paths_str[0], if_state_str,
-                                     if_id),
-            GrpcBase.mk_gnmi_if_path(self.list_paths_str[1]),
-            GrpcBase.mk_gnmi_if_path(self.list_paths_str[2].format(prefix_state_str))]
+                                     if_id)]
+        list_paths.extend([GrpcBase.mk_gnmi_if_path(self.list_paths_str[1]),
+            GrpcBase.mk_gnmi_if_path(self.list_paths_str[2].format(prefix_state_str))])
```
Comment (martin-volf) [inline tests/client_server_test_base.py:256]: Why don't you just add the two elements to the `list_path` definition above?

File: gnmi-tools.xml
[the remark this answers is not in the document]
Comment (micnovak) [reply gnmi-tools.xml:1]: To have small model with empty leaves and presence containters.

File: src/confd_gnmi_api_adapter.py [hunk 2]
```diff
@@ -522,10 +526,71 @@ def add_update_json(keypath, _value):
...
+                    else:
+                        elem = tr.get_elem(path_str)
+                        if csnode.info().defval() != elem:  # skip default values
+                            gnmi_value = self.make_gnmi_json_value(elem, csnode)
+                    if gnmi_value is not None:
+                        tr.pushd(path_str)
+                        kp = tr.getcwd_kpath()
+                        tr.popd()
+                        gnmi_path = self.make_gnmi_keypath(kp, csnode)
+                        updates.append(gnmi_pb2.Update(path=gnmi_path, val=gnmi_value))
```
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_api_adapter.py:550]: But why should it be exposed to outer scope, if not needed there?

File: src/confd_gnmi_api_adapter.py [hunk 3]
```diff
@@ -522,10 +526,71 @@ def add_update_json(keypath, _value):
...
+
+            if csnode.is_oper():
+                try:
+                    append_update_inner()
+                except _tm.error.Error:
+                    pass
+            else:
+                append_update_inner()
```
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_api_adapter.py:558]: Yes, just wanted to emphasize, there is incorrect behavior for oper data `tr.get_elem(path_str)` throws exception even though  `tr.exists(path_str)` returns `True` and it requires special handling.

File: src/confd_gnmi_api_adapter.py [hunk 4]
```diff
@@ -522,10 +526,71 @@ def add_update_json(keypath, _value):
...
+                if tr.exists(path_str):
+                    gnmi_value = None
+                    if csnode.is_empty_leaf():
+                        # See https://datatracker.ietf.org/doc/html/rfc7951#section-6.9
+                        gnmi_value = gnmi_pb2.TypedValue(json_ietf_val=json.dumps([None]).encode())
+                    elif csnode.is_p_container():
+                        gnmi_value = gnmi_pb2.TypedValue(json_ietf_val=json.dumps({}).encode())
+                    else:
+                        elem = tr.get_elem(path_str)
+                        if csnode.info().defval() != elem:  # skip default values
```
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_api_adapter.py:543]: Yes, we should review this. Our pytest tests expect only configured values to be returned.

File: gnmi-tools.yang
[the remark this answers is not in the document]
Comment (micnovak) [reply gnmi-tools.yang:1]: Reformatted to 2, added new line.

File: tests/client_server_test_base.py (hunk 1 above)
Comment (micnovak) [reply tests/client_server_test_base.py:256]: fixed

File: src/confd_gnmi_api_adapter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_api_adapter.py:550]: added generators, moved up to class scope

File: gnmi-tools.yang [hunk 5]
```diff
@@ -4,34 +4,34 @@ module gnmi-tools {
 
   description
     "This module contains a YANG data model for gnmi tools testing and
-      development.";
+                                                       development.";
```
Comment (martin-volf) [inline gnmi-tools.yang:7]: That auto-format did not go well...

**Review by martin-volf — APPROVED** [review]

File: test.sh
Comment (jomiklos) [inline test.sh:1]: end of line missing at the last line in file

File: gnmi-tools.yang (hunk 5 above)
Comment (micnovak) [reply gnmi-tools.yang]: fixed

File: test.sh
Comment (micnovak) [reply test.sh:1]: was already fixed

File: gnmi-tools.xml
[the remark this answers is not in the document]
Comment (micnovak) [reply gnmi-tools.xml:1]: it is data model for `gnmi-tools` testing

**Review by jomiklos — APPROVED** [review]

**Pull request merged**